File: src/main/java/net/jeikobu/jbase/command/CommandManager.java
package net.jeikobu.jbase.command;

import net.jeikobu.jbase.Localized;
import net.jeikobu.jbase.config.AbstractConfigManager;
import org.pmw.tinylog.Logger;
import sx.blah.discord.api.events.EventSubscriber;
import sx.blah.discord.handle.impl.events.guild.channel.message.MessageEvent;
import sx.blah.discord.handle.obj.IChannel;
import sx.blah.discord.handle.obj.IGuild;
import sx.blah.discord.handle.obj.IMessage;
import sx.blah.discord.handle.obj.IUser;

import javax.management.ReflectionException;
import java.lang.reflect.Constructor;
import java.lang.reflect.InvocationTargetException;
import java.util.*;

public class CommandManager extends Localized {
    private final AbstractConfigManager configManager;
    private final List<Class<? extends AbstractCommand>> registeredCommands;

    public CommandManager(AbstractConfigManager configManager) {
        this.configManager = configManager;
        registeredCommands = new ArrayList<>();
    }

    private static Command getCommandAnnotation(Class<? extends AbstractCommand> clazz) {
        Command c = clazz.getAnnotation(Command.class);
        if (c == null) {
            throw new IllegalArgumentException("This class is not correctly annotated!");
        } else {
            return c;
        }
    }

    private static AbstractCommand createCommandInstance(Class<? extends AbstractCommand> clazz, IGuild destGuild,
                                                         IChannel destChannel, IUser sender,
                                                         AbstractConfigManager configManager,
                                                         List<String> args) throws ReflectionException {
        try {
            Constructor<? extends AbstractCommand> constructor = clazz.getDeclaredConstructor(CommandData.class);

            return constructor.newInstance(new CommandData(destGuild, destChannel, sender, configManager, args));
        } catch (NoSuchMethodException | IllegalAccessException | InstantiationException | InvocationTargetException e) {
            throw new ReflectionException(e);
        }
    }

    public final void registerCommand(Class<? extends AbstractCommand> clazz) {
        registeredCommands.add(clazz);
    }

    public final void deregisterCommand(Class<? extends AbstractCommand> clazz) {
        registeredCommands.remove(clazz);
    }

    public final void registerAll(Collection<Class<? extends AbstractCommand>> classCollection) {
        registeredCommands.addAll(classCollection);
    }

    public final void deregisterAll(Collection<Class<? extends AbstractCommand>> classCollection) {
        registeredCommands.removeAll(classCollection);
    }

    @EventSubscriber
    public void onMessage(MessageEvent event) {
        IUser sender = event.getAuthor();
        IChannel destChannel = event.getChannel();
        IGuild destGuild = event.getGuild();

        String commandPrefix = configManager.getCommandPrefix(destGuild);
        Locale locale = configManager.getLocale(destGuild);

        String messageString = event.getMessage().getContent();
        IMessage message = event.getMessage();

        if (messageString.length() == 0 || !messageString.startsWith(commandPrefix)) {
            return;
        }

        List<String> args = Arrays.asList(messageString.split(" "));

        if (args.size() < 2) {
            return;
        }
        String suppliedCommandName = args.get(1);

        for (Class<? extends AbstractCommand> clazz : registeredCommands) {
            Command commandAnnotation = getCommandAnnotation(clazz);
            if (commandAnnotation.name().equals(suppliedCommandName)) {

                if (!sender.getPermissionsForGuild(destGuild)
                           .containsAll(Arrays.asList(commandAnnotation.permissions()))) {
                    destChannel.sendMessage(getLocalized(locale, "insufficientPermissions"));
                    return;
                }

                AbstractCommand command;

                try {
                    command = createCommandInstance(clazz, destGuild, destChannel, sender, configManager,
                                                    args.subList(2, args.size()));
                } catch (ReflectionException e) {
                    Logger.error(e);
                    destChannel.sendMessage(getLocalized(locale, "fatalError"));
                    return;
                }

                if (commandAnnotation.argsLength() + 2 > args.size()) {
                    try {
                        destChannel.sendMessage(command.usageMessage());
                    } catch (IllegalAccessException e) {
                        destChannel.sendMessage(getLocalized(locale, "notEnoughArgs"));
                    }
                    return;
                } else {
                    command.run(message);
                }
            }
        }

    }
}

File: src/main/java/net/jeikobu/kotomi/base/command/CommandManager.kt
package net.jeikobu.kotomi.base.command

import net.dv8tion.jda.core.events.Event
import net.dv8tion.jda.core.events.message.MessageReceivedEvent
import net.dv8tion.jda.core.hooks.EventListener
import net.jeikobu.kotomi.base.Localized
import net.jeikobu.kotomi.base.config.AbstractConfigManager
import org.pmw.tinylog.Logger

class CommandManager(private val configManager: AbstractConfigManager) : Localized(), EventListener {
    private val registeredCommands = ArrayList<AbstractCommand>()

    fun registerCommand(command: AbstractCommand) {
        registeredCommands.add(command)
    }

    fun deregisterCommand(command: AbstractCommand) {
        registeredCommands.remove(command)
    }

    fun registerAll(cmdCollection: Collection<AbstractCommand>) {
        registeredCommands.addAll(cmdCollection)
    }

    fun deregisterAll(cmdCollection: Collection<AbstractCommand>) {
        registeredCommands.removeAll(cmdCollection)
    }

    override fun onEvent(e: Event) {
        if (e is MessageReceivedEvent) {
            val sender = e.guild.getMember(e.author)
            val commandPrefix = configManager.getCommandPrefix(e.guild)
            val locale = configManager.getLocale(e.guild)
            val message = e.message
            val messageStr = message.contentRaw

            if (messageStr.isEmpty() || !messageStr.startsWith(commandPrefix)) {
                return
            }

            val args = messageStr.split(" ")
            if (args.size < 2) {
                return
            }

            val matchingCommands = registeredCommands.filter { it.name == args[1] || it.aliases.contains(args[1]) }
            if (matchingCommands.size > 1) {
                Logger.error("Critical error: More than one command matching!")
                e.textChannel.sendMessage(getLocalized(locale, "fatalError", "More than one command matching")).queue()
            } else if (matchingCommands.isEmpty()) {
                return
            }

            val command = matchingCommands.first()
            val commandData = CommandData(e.guild, e.textChannel, sender, configManager, args.subList(2, args.size))

            if (!sender.permissions.containsAll(command.permissions)) {
                e.textChannel.sendMessage(getLocalized(locale, "insufficientPermissions")).queue()
                return
            }

            if (command.argsLength + 2 > args.size) {
                e.textChannel.sendMessage(getLocalized(locale, "notEnoughArgs")).queue()
                return
            } else {
                command.run(commandData, message)
            }
        }
    }
}